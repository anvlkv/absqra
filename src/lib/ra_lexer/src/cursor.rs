use super::*;
use core::iter::Peekable;
use std::str::Chars;
use std::convert::{TryInto, TryFrom};

pub (crate) struct Cursor<'char> {
    pub position: Position,
    pub level: u16,
    pub ch: Option<char>,
    pub is_line_start: bool,
    pub ruler: Option<u16>,
    indent_width: u8,
    chars: Peekable<Chars<'char>>,
    initial_level: u16
}

pub (crate) const EOF_CHAR: char = '\0';

/// True if `c` is considered a whitespace
pub (crate) fn is_whitespace(c: &char) -> bool {
    match c {
        // Usual ASCII suspects
        | '\u{0009}' // \t
        | '\u{000B}' // vertical tab
        | '\u{000C}' // form feed
        | '\u{0020}' // space

        // Bidi markers
        | '\u{200E}' // LEFT-TO-RIGHT MARK
        | '\u{200F}' // RIGHT-TO-LEFT MARK
            => true,
        _ => false,
    }
}

/// True if `c` is considered an end of line
pub (crate) fn is_end_of_line(c: &char) -> bool {
    match c {
        // Usual ASCII suspects
        | '\u{000A}' // \n
        | '\u{000D}' // \r

        // NEXT LINE from latin1
        | '\u{0085}'

        // // Dedicated whitespace characters from Unicode
        | '\u{2028}' // LINE SEPARATOR
        | '\u{2029}' // PARAGRAPH SEPARATOR
            => true,
        _ => false,
    }
}

impl <'c> Cursor<'c> {
    pub fn bump(&mut self) {
        if let Some(ch) = self.ch {
            match ch {
                c if is_end_of_line(&c) => {
                    self.level();
                },
                _ => {
                    match self.chars.next() {
                        Some(ch) => {
                            self.ch = Some(ch);
                            self.position.1 += 1;
                            self.is_line_start = false;
                        }
                        None => {
                            self.ch = None;
                        }
                    }
                }
            }
        }
    }

    fn level(&mut self) {
        assert!(self.ch.is_some() && is_end_of_line(&self.ch.unwrap()));
        self.position.0 += 1;
        self.position.1 = 0;

        match self.first_ahead() {
            EOF_CHAR => {
                self.ch = None;
            },
            mut f_ch if is_whitespace(&f_ch) => {
                let mut new_indent_width = None;
                let mut new_level = 0;
                let white_space_ch = f_ch.clone();
                let limit = match self.ruler {
                    Some(l) => l,
                    None => std::u16::MAX
                };
                
                while is_whitespace(&f_ch) && white_space_ch == f_ch {
                    self.position.1 += 1;
                    self.ch = self.chars.next();
                    
                    
                    if self.indent_width == 0 {
                        new_indent_width = Some((self.position.1).try_into().unwrap());
                    }
                    else {
                        let indent_16 = u16::from(self.indent_width);
                        if self.position.1 % indent_16  == 0 {
                            new_level = self.position.1 / indent_16;
                        }
                    }

                    f_ch = self.first_ahead();

                    if new_level >= limit {
                        break;
                    }
                }

                self.position.1 += 1;
                self.ch = self.chars.next();

                match new_indent_width {
                    Some(width) => {
                        self.indent_width = width;
                        self.level = 2;
                    },
                    None => {
                        self.level = new_level + 1;
                    }
                }
            },
            c if is_end_of_line(&c) => {
                self.ch = self.chars.next();
                self.level();
            },
            _ => {
                self.level = self.initial_level;
                self.ch = self.chars.next();
                self.position.1 += 1;
            }
        }
       
        self.is_line_start = true;
    }

    pub fn first_ahead(&mut self) -> char {
        self.chars.peek().unwrap_or(&EOF_CHAR).clone()
    }
}

impl <'c> From<& 'c str> for Cursor<'c> {
    fn from(input: & 'c str) -> Self {
        let input_c = input.clone(); 
        let mut chars = input_c.chars().peekable();
        Self {
            level: 1,
            ch: chars.next(),
            position: Position(1, 1),
            is_line_start: true,
            ruler: None,
            indent_width: 0,
            chars,
            initial_level: 1,
        }
    }
}

impl <'c> Iterator for Cursor<'c> {
    type Item = Result<RaToken, LexerError>;

    fn next(&mut self) -> Option<Self::Item> {
        match self.ch {
            Some(ch) => match ch {
                ch if is_whitespace(&ch) || is_end_of_line(&ch) => {
                    self.bump();
                    self.next()
                },
                _ => Some(RaToken::try_from(self))
            },
            None => None
        }
    }
}