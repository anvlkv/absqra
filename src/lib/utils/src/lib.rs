use std::path::Path;
use std::fs::{DirEntry, read_dir};

pub fn files_from_dir_recursively(path: &str) -> Vec<DirEntry> {
    read_dir(Path::new(path)).unwrap().map(|f| f.unwrap()).flat_map(|entry| {
        match entry.metadata().unwrap() {
            meta if meta.is_dir() => files_from_dir_recursively(&entry.path().into_os_string().into_string().unwrap()),
            meta if meta.is_file() => vec![entry],
            _ => panic!("unsupported metadata")
        }
    }).collect()
}