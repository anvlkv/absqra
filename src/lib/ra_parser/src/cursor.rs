use ra_lexer::token::{Token};
use ra_lexer::cursor::Position;
use core::iter::Peekable;

pub(crate) struct Cursor<'token, I> 
    where I: Iterator<Item = Token<'token>> 
{
    pub position: Position,
    pub level: u16,
    tokens: Peekable<I>,
    consumed_len: usize
}

impl <'token, I> Cursor<'token, I> 
where I: Iterator<Item = Token<'token>> 
{
    pub fn new(tokens: I) -> Self
        
    {
        Self {
            position: Position(0, 0),
            level: 0,
            tokens: tokens.peekable(),
            consumed_len: 0
        }
    }

    pub fn first_ahead(&mut self) -> Option<&Token<'token>> {
        self.tokens.peek()
    }

    pub fn bump(&mut self) -> Option<Token<'token>> {
        let tok = self.tokens.next();
        match tok {
            Some(token) => {
                self.position = token.position.1;
                self.level = token.level;
                self.consumed_len += 1;
                Some(token)
            }
            None => None
        }
    }

    pub fn is_eof(& mut self) -> bool {
        self.first_ahead() == None
    }

    // pub fn do_while<F, D>(&mut self, mut test: F, mut cb: D)
    //     where F: FnMut (u16, Position, TokenKind) -> bool,
    //     D: FnMut (Token)
    // {
    //     loop {
    //         let next_token = self.first_ahead();
    //         match next_token {
    //             Some(t) => {
    //                 if test(t.level, t.position.0, t.kind) {
    //                     cb(self.bump().unwrap())
    //                 }
    //                 else {
    //                     break
    //                 }
    //             },
    //             None => break
    //         }
    //     }
    // }

    // pub fn read_within(&mut self, open: TokenKind, close: TokenKind) -> Vec<Token> {
    //     let mut read: Vec<Token> = Vec::new();

    //     loop {
    //         let next_token = self.first_ahead();

    //         match next_token {
    //             Some(tok) => {
    //                 match tok.kind {
    //                     k if k == open => {
    //                         let open_token = self.bump().unwrap();
    //                         let mut nested = self.read_within(open_token.kind, close.clone());
    //                         read.append(&mut nested);
    //                     },
    //                     k if k == close => {
    //                         self.bump();
    //                         break
    //                     },
    //                     _ => read.push(self.bump().unwrap())
    //                 }
    //             },
    //             None => break
    //         }

    //         // println!("{:?}", &read);
    //     }

    //     read
    // }
}