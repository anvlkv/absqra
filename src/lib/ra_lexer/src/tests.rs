use crate::cursor::{Cursor, Position};
use crate::{tokenize, Token, TokenKind};

mod cursor {
    use super::{Cursor, Position};
    #[test]
    fn it_should_create() {
        Cursor::new("abc", Position(1, 0), 0, 0);
    }

    #[test]
    fn it_should_give_first_char() {
        let cur = Cursor::new("abc", Position(1, 0), 0, 0);
        assert_eq!(cur.first_ahead(), 'a');
    }

    #[test]
    fn it_should_give_second_char() {
        let cur = Cursor::new("abc", Position(1, 0), 0, 0);
        assert_eq!(cur.second_ahead(), 'b');
    }

    #[test]
    fn it_should_return_next_char() {
        let mut cur = Cursor::new("a", Position(1, 0), 0, 0);
        assert_eq!(cur.bump().unwrap(), 'a');
    }

    #[test]
    fn it_should_return_none_at_end_of_input() {
        let mut cur = Cursor::new("a", Position(1, 0), 0, 0);
        cur.bump();
        assert_eq!(cur.bump(), None);
    }

    #[test]
    fn it_should_confirm_eof() {
        let mut cur = Cursor::new("a", Position(1, 0), 0, 0);
        cur.bump();
        cur.bump();
        cur.bump();
        assert_eq!(cur.is_eof(), true);
    }

    #[test]
    fn it_should_return_amount_of_consumed_symbols() {
        let mut cur = Cursor::new("abc", Position(1, 0), 0, 0);
        cur.bump();
        cur.bump();
        cur.bump();
        assert_eq!(cur.len_consumed(), 3);
    }

    #[test]
    fn it_should_not_panic_when_encountering_funny_characters() {
        let mut cur = Cursor::new("ðŸš¬", Position(1, 0), 0, 0);
        assert_eq!(cur.bump().unwrap(), 'ðŸš¬');
    }

    #[test]
    fn it_should_track_position() {
        let mut cur = Cursor::new("abc\nSOME", Position(1, 0), 0, 0);
        assert_eq!(cur.bump().unwrap(), 'a');
        assert_eq!(cur.position, Position(1, 1));
        assert_eq!(cur.bump().unwrap(), 'b');
        assert_eq!(cur.position, Position(1, 2));
        assert_eq!(cur.bump().unwrap(), 'c');
        assert_eq!(cur.position, Position(1, 3));
        assert_eq!(cur.bump().unwrap(), 'S');
        assert_eq!(cur.position, Position(2, 1));
        assert_eq!(cur.bump().unwrap(), 'O');
        assert_eq!(cur.position, Position(2, 2));
        assert_eq!(cur.bump().unwrap(), 'M');
        assert_eq!(cur.position, Position(2, 3));
    }

    #[test]
    fn it_should_track_indent_level() {
        let mut cur = Cursor::new("a\n\tb\n\t\tc\nd", Position(1, 0), 0, 0);
        assert_eq!(cur.bump().unwrap(), 'a');
        assert_eq!(cur.level, 0);
        assert_eq!(cur.bump().unwrap(), 'b');
        assert_eq!(cur.level, 1);
        assert_eq!(cur.bump().unwrap(), 'c');
        assert_eq!(cur.level, 2);
        assert_eq!(cur.position, Position(3, 3));
        assert_eq!(cur.bump().unwrap(), 'd');
        assert_eq!(cur.level, 0);
    }
}

mod lib {
    use super::{tokenize, Position, Token, TokenKind};
    

    #[test]
    fn it_should_create_iterator_of_tokens() {
        let mut stream = tokenize("abc");
        stream.next();
    }

    #[test]
    fn it_should_parse_multiple_lines() {
        let mut stream = tokenize("!?&/\n\tabc");
        stream.next();
        stream.next();
        stream.next();
        stream.next();
        assert_eq!(
            stream.next().unwrap(),
            Token {
                kind: Some(TokenKind::Identifier("abc")),
                content: "abc",
                position: (Position(2, 1), Position(2, 4)),
                len: 3,
                level: 1,
            }
        )
    }

    #[test]
    fn it_should_parse_identifiers() {
        let mut stream = tokenize("abc");
        assert_eq!(
            stream.next().unwrap(),
            Token {
                kind: Some(TokenKind::Identifier("abc")),
                content: "abc",
                position: (Position(1, 0), Position(1, 3)),
                len: 3,
                level: 0
            }
        );
    }

    use ra_common::insta::assert_debug_snapshot;
    use utils::for_each_ra_example_file;
    use std::fs::File;
    use std::io::Read;

    #[test]
    fn it_should_match_snapshots() {
        for_each_ra_example_file(|example| {
                let mut file = File::open(example.path()).unwrap();
    
                let mut contents = String::new();
                file.read_to_string(&mut contents).unwrap();

                println!("{}", example.file_name().to_str().unwrap());

                let tokens: Vec<Token> = tokenize(&contents).collect();
                assert_debug_snapshot!(example.path().to_str().unwrap(), tokens)
        });
    }

    mod strings {
        use super::{tokenize, Position, Token, TokenKind};
        #[test]
        fn it_should_parse_string_literals_with_double_quotes() {
            let mut stream = tokenize("\"some\"");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::StringLiteral("some")),
                    content: "some",
                    len: 6,
                    position: (Position(1, 0), Position(1, 6)),
                    ..Default::default()
                }
            )
        }

        #[test]
        fn it_should_parse_string_literals_with_single_quotes() {
            let mut stream = tokenize("\'some\'");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::StringLiteral("some")),
                    content: "some",
                    len: 6,
                    position: (Position(1, 0), Position(1, 6)),
                    ..Default::default()
                }
            )
        }
    }
    mod comments {
        use super::{tokenize, Position, Token, TokenKind};
        #[test]
        fn it_should_parse_single_line_comments() {
            let mut stream = tokenize("//abc");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Comment),
                    content: "abc",
                    position: (Position(1, 0), Position(1, 5)),
                    len: 5,
                    level: 0
                }
            );
        }
        #[test]
        fn it_should_parse_multi_line_comments() {
            let mut stream = tokenize("/*abc\nSOME*/");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Comment),
                    content: "abc\nSOME",
                    position: (Position(1, 0), Position(2, 6)),
                    len: 12,
                    level: 0
                }
            );
        }

        #[test]
        fn it_should_parse_multi_line_comments_with_fringe() {
            let mut stream = tokenize("/*\n* abc\n* SOME\n*/");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Comment),
                    content: "\n* abc\n* SOME\n",
                    position: (Position(1, 0), Position(4, 2)),
                    len: 18,
                    level: 0
                }
            );
        }

        #[test]
        fn it_should_parse_comments_followed_by_one_another() {
            let mut stream = tokenize("//some\n\n/*\n* abc\n* SOME\n*/");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Comment),
                    content: "some",
                    position: (Position(1, 0), Position(1, 6)),
                    len: 6,
                    level: 0
                }
            );

            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Comment),
                    content: "\n* abc\n* SOME\n",
                    position: (Position(3, 0), Position(6, 2)),
                    len: 18,
                    level: 0
                }
            );
        }
    }
    
    mod numbers {
        use super::{tokenize, Position, Token, TokenKind};
        #[test]
        fn it_should_parse_numbers() {
            let mut stream = tokenize("123");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Int(123)),
                    content: "123",
                    position: (Position(1, 0), Position(1, 3)),
                    len: 3,
                    level: 0
                }
            );
        }

        #[test]
        fn it_should_parse_numbers_with_decimal_separator() {
            let mut stream = tokenize("123,321");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Float(123.321)),
                    content: "123,321",
                    position: (Position(1, 0), Position(1, 7)),
                    len: 7,
                    level: 0
                }
            );
        }

        #[test]
        fn it_should_parse_numbers_with_another_decimal_separator() {
            let mut stream = tokenize("123.321");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Float(123.321)),
                    content: "123.321",
                    position: (Position(1, 0), Position(1, 7)),
                    len: 7,
                    level: 0
                }
            );
        }

        #[test]
        fn it_should_parse_numbers_with_decimal_and_thousands_separator() {
            let mut stream = tokenize("123.321,456");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Float(123321.456)),
                    content: "123.321,456",
                    position: (Position(1, 0), Position(1, 11)),
                    len: 11,
                    level: 0
                }
            );
        }

        #[test]
        fn it_should_parse_numbers_with_decimal_and_multiple_thousands_separator() {
            let mut stream = tokenize("123.321.123,456");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Float(123321123.456)),
                    content: "123.321.123,456",
                    position: (Position(1, 0), Position(1, 15)),
                    len: 15,
                    level: 0
                }
            );
        }

        #[test]
        fn it_should_parse_numbers_with_another_decimal_and_multiple_thousands_separators() {
            let mut stream = tokenize("123,321,123.456");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Float(123321123.456)),
                    content: "123,321,123.456",
                    position: (Position(1, 0), Position(1, 15)),
                    len: 15,
                    level: 0
                }
            );
        }

        #[test]
        #[should_panic]
        fn it_should_panic_when_encountering_multiple_decimal_separators() {
            let mut stream = tokenize("123.321.123,456,654");
            stream.next();
        }

        #[test]
        fn it_should_parse_negative_integer() {
            let mut stream = tokenize("-123");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Int(-123)),
                    content: "-123",
                    position: (Position(1, 0), Position(1, 4)),
                    len: 4,
                    level: 0
                }
            )
        }

        #[test]
        fn it_should_parse_negative_float() {
            let mut stream = tokenize("-123.312");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Float(-123.312)),
                    content: "-123.312",
                    position: (Position(1, 0), Position(1, 8)),
                    len: 8,
                    level: 0
                }
            )
        }
    }

    mod content_block {
        use super::{tokenize, Position, Token, TokenKind};
        #[test]
        fn it_should_parse_content_blocks() {
            let mut stream = tokenize("`abc`");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::ContentBlock),
                    content: "abc",
                    position: (Position(1, 0), Position(1, 5)),
                    len: 5,
                    level: 0
                }
            );
        }

        #[test]
        fn it_should_keep_inner_indents_when_parsing_content() {
            let mut stream = tokenize("`\n\tabc\n\t\tabc\n`");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::ContentBlock),
                    content: "\n\tabc\n\t\tabc\n",
                    position: (Position(1, 0), Position(4, 1)),
                    len: 14,
                    level: 0
                }
            );
        }

        #[test]
        #[should_panic]
        fn it_should_panic_when_content_block_is_not_closed_properly() {
            let mut stream = tokenize("`\n\tabc`");
            stream.next();
        }
    }

    mod single_char_tokens {
        use super::{tokenize, Position, Token, TokenKind};
        #[test]
        fn it_should_parse_tokens() {
            let mut stream = tokenize("!?&/#");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Exclamation),
                    position: (Position(1, 0), Position(1, 1)),
                    content: "!",
                    level: 0,
                    ..Default::default()
                }
            );
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Question),
                    position: (Position(1, 1), Position(1, 2)),
                    content: "?",
                    level: 0,
                    ..Default::default()
                }
            );
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Ampersand),
                    position: (Position(1, 2), Position(1, 3)),
                    content: "&",
                    level: 0,
                    ..Default::default()
                }
            );
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Slash),
                    position: (Position(1, 3), Position(1, 4)),
                    content: "/",
                    level: 0,
                    ..Default::default()
                }
            );
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::HashPound),
                    position: (Position(1, 4), Position(1, 5)),
                    content: "#",
                    level: 0,
                    ..Default::default()
                }
            );
        }
    }


}
