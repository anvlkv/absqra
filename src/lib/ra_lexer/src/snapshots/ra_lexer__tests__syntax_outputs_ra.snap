---
source: src/lib/ra_lexer/src/tests.rs
expression: tokens
---
[
  {
    "kind": {
      "Identifier": "single_variable"
    },
    "position": [
      [
        1,
        1
      ],
      [
        1,
        16
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Comment": [
        " just an empty variable",
        false
      ]
    },
    "position": [
      [
        1,
        17
      ],
      [
        1,
        42
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "IntegerLiteral": 1
    },
    "position": [
      [
        3,
        1
      ],
      [
        0,
        0
      ]
    ],
    "level": 1
  },
  {
    "kind": "Plus",
    "position": [
      [
        3,
        3
      ],
      [
        3,
        4
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "IntegerLiteral": 1
    },
    "position": [
      [
        3,
        5
      ],
      [
        0,
        0
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Comment": [
        " expression that has an output of 2",
        false
      ]
    },
    "position": [
      [
        3,
        7
      ],
      [
        3,
        44
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "assignment"
    },
    "position": [
      [
        5,
        1
      ],
      [
        5,
        11
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Comment": [
        " expression that has value 1 assigned",
        false
      ]
    },
    "position": [
      [
        5,
        12
      ],
      [
        5,
        51
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "IntegerLiteral": 1
    },
    "position": [
      [
        6,
        5
      ],
      [
        0,
        0
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "assignment"
    },
    "position": [
      [
        8,
        1
      ],
      [
        8,
        11
      ]
    ],
    "level": 1
  },
  {
    "kind": "Plus",
    "position": [
      [
        8,
        12
      ],
      [
        8,
        13
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "IntegerLiteral": 2
    },
    "position": [
      [
        8,
        14
      ],
      [
        0,
        0
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Comment": [
        " expression that has an output of 3",
        false
      ]
    },
    "position": [
      [
        8,
        16
      ],
      [
        8,
        53
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "nested_variable"
    },
    "position": [
      [
        10,
        1
      ],
      [
        10,
        16
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Comment": [
        " expression that has output assigned",
        false
      ]
    },
    "position": [
      [
        10,
        17
      ],
      [
        10,
        55
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "assignment"
    },
    "position": [
      [
        11,
        5
      ],
      [
        11,
        15
      ]
    ],
    "level": 2
  },
  {
    "kind": "Plus",
    "position": [
      [
        11,
        16
      ],
      [
        11,
        17
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "IntegerLiteral": 2
    },
    "position": [
      [
        11,
        18
      ],
      [
        0,
        0
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Comment": [
        " expression that has an output of 3",
        false
      ]
    },
    "position": [
      [
        11,
        20
      ],
      [
        11,
        57
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "assignment"
    },
    "position": [
      [
        13,
        1
      ],
      [
        13,
        11
      ]
    ],
    "level": 1
  },
  {
    "kind": "Pipe",
    "position": [
      [
        13,
        12
      ],
      [
        13,
        13
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "nested_variable"
    },
    "position": [
      [
        13,
        14
      ],
      [
        13,
        29
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Comment": [
        " logical expression OR",
        false
      ]
    },
    "position": [
      [
        13,
        30
      ],
      [
        13,
        54
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "assignment"
    },
    "position": [
      [
        15,
        1
      ],
      [
        15,
        11
      ]
    ],
    "level": 1
  },
  {
    "kind": "Greater",
    "position": [
      [
        15,
        12
      ],
      [
        15,
        13
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "nested_variable"
    },
    "position": [
      [
        15,
        14
      ],
      [
        15,
        29
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Comment": [
        " logical expression Comparison GT",
        false
      ]
    },
    "position": [
      [
        15,
        30
      ],
      [
        15,
        65
      ]
    ],
    "level": 1
  },
  {
    "kind": "OpenParentheses",
    "position": [
      [
        17,
        1
      ],
      [
        17,
        2
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "IntegerLiteral": 3
    },
    "position": [
      [
        17,
        2
      ],
      [
        0,
        0
      ]
    ],
    "level": 1
  },
  {
    "kind": "Asterisk",
    "position": [
      [
        17,
        3
      ],
      [
        17,
        4
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "IntegerLiteral": 5
    },
    "position": [
      [
        17,
        4
      ],
      [
        0,
        0
      ]
    ],
    "level": 1
  },
  {
    "kind": "CloseParentheses",
    "position": [
      [
        17,
        5
      ],
      [
        17,
        6
      ]
    ],
    "level": 1
  },
  {
    "kind": "Plus",
    "position": [
      [
        17,
        7
      ],
      [
        17,
        8
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "IntegerLiteral": 22
    },
    "position": [
      [
        17,
        9
      ],
      [
        17,
        11
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Comment": [
        " grouped expression that has an output of 37",
        false
      ]
    },
    "position": [
      [
        17,
        12
      ],
      [
        17,
        58
      ]
    ],
    "level": 1
  },
  {
    "kind": "OpenParentheses",
    "position": [
      [
        19,
        1
      ],
      [
        19,
        2
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "assignment"
    },
    "position": [
      [
        19,
        2
      ],
      [
        19,
        12
      ]
    ],
    "level": 1
  },
  {
    "kind": "Pipe",
    "position": [
      [
        19,
        13
      ],
      [
        19,
        14
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "nested_variable"
    },
    "position": [
      [
        19,
        15
      ],
      [
        19,
        30
      ]
    ],
    "level": 1
  },
  {
    "kind": "CloseParentheses",
    "position": [
      [
        19,
        30
      ],
      [
        19,
        31
      ]
    ],
    "level": 1
  },
  {
    "kind": "Plus",
    "position": [
      [
        19,
        32
      ],
      [
        19,
        33
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "IntegerLiteral": 3
    },
    "position": [
      [
        19,
        34
      ],
      [
        0,
        0
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Comment": [
        " grouped expression with logical expression, has an output of 6",
        false
      ]
    },
    "position": [
      [
        19,
        36
      ],
      [
        19,
        101
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "IntegerLiteral": 1
    },
    "position": [
      [
        21,
        1
      ],
      [
        0,
        0
      ]
    ],
    "level": 1
  },
  {
    "kind": "Plus",
    "position": [
      [
        21,
        3
      ],
      [
        21,
        4
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "IntegerLiteral": 1
    },
    "position": [
      [
        21,
        5
      ],
      [
        0,
        0
      ]
    ],
    "level": 1
  },
  {
    "kind": "Plus",
    "position": [
      [
        21,
        7
      ],
      [
        21,
        8
      ]
    ],
    "level": 1
  },
  {
    "kind": "OpenParentheses",
    "position": [
      [
        21,
        9
      ],
      [
        21,
        10
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "IntegerLiteral": 2
    },
    "position": [
      [
        21,
        10
      ],
      [
        0,
        0
      ]
    ],
    "level": 1
  },
  {
    "kind": "Asterisk",
    "position": [
      [
        21,
        11
      ],
      [
        21,
        12
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "IntegerLiteral": 2
    },
    "position": [
      [
        21,
        12
      ],
      [
        0,
        0
      ]
    ],
    "level": 1
  },
  {
    "kind": "CloseParentheses",
    "position": [
      [
        21,
        13
      ],
      [
        21,
        14
      ]
    ],
    "level": 1
  },
  {
    "kind": "Plus",
    "position": [
      [
        21,
        15
      ],
      [
        21,
        16
      ]
    ],
    "level": 1
  },
  {
    "kind": "OpenParentheses",
    "position": [
      [
        21,
        17
      ],
      [
        21,
        18
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "IntegerLiteral": 3
    },
    "position": [
      [
        21,
        18
      ],
      [
        0,
        0
      ]
    ],
    "level": 1
  },
  {
    "kind": "Asterisk",
    "position": [
      [
        21,
        19
      ],
      [
        21,
        20
      ]
    ],
    "level": 1
  },
  {
    "kind": "OpenParentheses",
    "position": [
      [
        21,
        20
      ],
      [
        21,
        21
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "IntegerLiteral": 2
    },
    "position": [
      [
        21,
        21
      ],
      [
        0,
        0
      ]
    ],
    "level": 1
  },
  {
    "kind": "Plus",
    "position": [
      [
        21,
        22
      ],
      [
        21,
        23
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "IntegerLiteral": 2
    },
    "position": [
      [
        21,
        23
      ],
      [
        0,
        0
      ]
    ],
    "level": 1
  },
  {
    "kind": "CloseParentheses",
    "position": [
      [
        21,
        24
      ],
      [
        21,
        25
      ]
    ],
    "level": 1
  },
  {
    "kind": "CloseParentheses",
    "position": [
      [
        21,
        25
      ],
      [
        21,
        26
      ]
    ],
    "level": 1
  },
  {
    "kind": "Minus",
    "position": [
      [
        21,
        27
      ],
      [
        21,
        28
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "IntegerLiteral": 18
    },
    "position": [
      [
        21,
        29
      ],
      [
        21,
        31
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Comment": [
        " bit more grouping, has an output of 0",
        false
      ]
    },
    "position": [
      [
        21,
        32
      ],
      [
        21,
        72
      ]
    ],
    "level": 1
  }
]
